var/spack/repos/builtin/packages/aws-parallelcluster/package.py

```diff
@@ -11,9 +11,9 @@ class AwsParallelcluster(PythonPackage):
     tool to deploy and manage HPC clusters in the AWS cloud."""
 
     homepage = "https://github.com/aws/aws-parallelcluster"
-    url      = "https://pypi.io/packages/source/a/aws-parallelcluster/aws-parallelcluster-2.5.0.tar.gz"
+    url      = "https://pypi.io/packages/source/a/aws-parallelcluster/aws-parallelcluster-2.5.1.tar.gz"
 
-    version('2.5.0', sha256='3b0209342ea0d9d8cc95505456103ad87c2d4e35771aa838765918194efd0ad3')
+    version('2.5.1', sha256='4fd6e14583f8cf81f9e4aa1d6188e3708d3d14e6ae252de0a94caaf58be76303')
```

Comment:
In Spack we like to keep the old versions for reproducibility. Can you keep 2.5.0 and 2.5.1?

Reply:
@adamjstewart Added collaborators and old version back! 

How does it create the URL for the previous versions? That seems wrong

Reply:
Magic.

Just kidding. Spack has really clever logic for extrapolating the download URL. For example, you can run:
```console
$ spack url parse https://pypi.io/packages/source/a/aws-parallelcluster/aws-parallelcluster-2.5.1.tar.gz
==> Parsing URL: https://pypi.io/packages/source/a/aws-parallelcluster/aws-parallelcluster-2.5.1.tar.gz

==> Matched version regex  0: r'^[a-zA-Z+._-]+[._-]v?(\\d[\\d._-]*)$'
==> Matched  name   regex  4: r'pypi\\.(?:python\\.org|io)/packages/source/[A-Za-z\\d]/([^/]+)'

==> Detected:
    https://pypi.io/packages/source/a/aws-parallelcluster/aws-parallelcluster-2.5.1.tar.gz
                                      -------------------                     ~~~~~
    name:    aws-parallelcluster
    version: 2.5.1

==> Substituting version 9.9.9b:
    https://pypi.io/packages/source/a/aws-parallelcluster/aws-parallelcluster-9.9.9b.tar.gz
                                      -------------------                     ~~~~~~
```
That shows you that Spack does indeed correctly detect the version in the URL and substitute it properly. This allows packages to contain hundreds of versions but only a single URL. More complicated projects that change their URL can override this on a version-specific level, or define a `url_for_version` to programmatically determine the URL.

Reply:
This is dark magic 🧙🏻‍♂️